tests/tests_full_scan_engine_udf.py

```diff
@@ -0,0 +1,355 @@
+        if service.service.name == 'objects00':
+          inference_service = service
+          outputs = await inference_service.infer(blob_id)
+          return outputs[0]
+
+
+    def objects_inference(blob_id):
+      df = pd.DataFrame({'blob_id': [blob_id]})
+      return asyncio_run(async_objects_inference(df))
+
+
+    async def async_lights_inference(blob_id):
+      # Simulate some async operation
+      for service in aidb_engine._config.inference_bindings:
+        if service.service.name == 'lights01':
+          inference_service = service
+          outputs = await inference_service.infer(blob_id)
+          return outputs[0]
+
+
+    def lights_inference(blob_id):
+      df = pd.DataFrame({'blob_id': [blob_id]})
+      return asyncio_run(async_lights_inference(df))
+
+
+    async def async_colors_inference(blob_id):
+      # Simulate some async operation
+      for service in aidb_engine._config.inference_bindings:
+        if service.service.name == 'colors02':
+          inference_service = service
+          outputs = await inference_service.infer(blob_id)
+          return outputs[0]
+
+
+    def colors_inference(blob_id, object_id):
+      df = pd.DataFrame({'blob_id': [blob_id], 'input_col2': object_id})
+      return asyncio_run(async_colors_inference(df))
+
+
+    async def async_counts_inference(blob_id):
+      # Simulate some async operation
+      for service in aidb_engine._config.inference_bindings:
+        if service.service.name == 'counts03':
+          inference_service = service
+          outputs = await inference_service.infer(blob_id)
+          return outputs[0]
+
+
+    def counts_inference(blob_id):
+      df = pd.DataFrame({'blob_id': [blob_id]})
+      return asyncio_run(async_counts_inference(df))
+
+
+    aidb_engine._config.add_user_defined_function('sum_function', sum_function)
+    aidb_engine._config.add_user_defined_function('is_equal', is_equal)
+    aidb_engine._config.add_user_defined_function('max_function', max_function)
+    aidb_engine._config.add_user_defined_function('power_function', power_function)
+    aidb_engine._config.add_user_defined_function('replace_color', replace_color)
+    aidb_engine._config.add_user_defined_function('objects_inference', objects_inference)
+    aidb_engine._config.add_user_defined_function('lights_inference', lights_inference)
+    aidb_engine._config.add_user_defined_function('colors_inference', colors_inference)
+    aidb_engine._config.add_user_defined_function('counts_inference', counts_inference)
+
+
+
+  async def test_jackson_number_objects(self):
+
+    dirname = os.path.dirname(__file__)
+    data_dir = os.path.join(dirname, 'data/jackson')
+
+    p = Process(target=run_server, args=[str(data_dir)])
+    p.start()
+    time.sleep(3)
+    gt_engine, aidb_engine = await setup_gt_and_aidb_engine(DB_URL, data_dir)
+
+    register_inference_services(aidb_engine, data_dir)
+
+    self.add_user_defined_function(aidb_engine)
+
+    queries = [
```

Comment:
Given this is so complicated, we should document the UDF behavior somewhere

Reply:
https://github.com/ddkang/aidb/wiki/User-Defined-Function-Usage
https://github.com/ddkang/aidb/wiki/User-Defined-Function-%E2%80%90-Design-Document